File: backend/pkg/domain/model/bookrating.go
package model

type BookRating struct {
	bookID BookID
	userID UserID
	value  int
}

func NewBookRating(
	bookID BookID,
	userID UserID,
	value int,
) BookRating {
	return BookRating{
		bookID: bookID,
		userID: userID,
		value:  value,
	}
}

func (bookRating *BookRating) BookID() BookID {
	return bookRating.bookID
}

func (bookRating *BookRating) UserID() UserID {
	return bookRating.userID
}

func (bookRating *BookRating) Value() int {
	return bookRating.value
}

File: frontend/src/api/bookRating.ts
import { UpdateBookRating, GetBookRatingResponse } from "@shared/types/bookRating";
import { handleApiError } from "./utils/handleApiError";

export const bookRatingApi = {
  PREFIX: "/api/v1/book",

  updateRating(bookId: string, body: UpdateBookRating): Promise<unknown> {
    return fetch(`${this.PREFIX}/${bookId}/raiting`, {
      method: "POST",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    }).then((res) => {
      if (!res.ok) {
        return handleApiError(res, () => this.updateRating(bookId, body));
      }
      return res.json();
    });
  },

  deleteRating(bookId: string): Promise<unknown> {
    return fetch(`${this.PREFIX}/${bookId}/raiting`, {
      method: "DELETE",
      credentials: "include",
    }).then((res) => {
      if (!res.ok) {
        return handleApiError(res, () => this.deleteRating(bookId));
      }
      return res.json();
    });
  },

  getRating(bookId: string): Promise<GetBookRatingResponse> {
    return fetch(`${this.PREFIX}/${bookId}/raiting`, {
      method: "GET",
      credentials: "include",
    }).then((res) => {
      if (!res.ok) {
        return handleApiError(res, () => this.getRating(bookId));
      }
      return res.json();
    });
  },
} as const;
